File: backend/test_e2e.py
import requests
import time
import json
import os

BASE_URL = "http://127.0.0.1:8001"

def run_test():
    print("Starting E2E API Test...")
    
    # 1. Super Admin creates patient
    admin_login = requests.post(f"{BASE_URL}/token", data={"username": "admin@medflow.com", "password": "admin123"})
    admin_token = admin_login.json().get("access_token")
    headers = {"Authorization": f"Bearer {admin_token}"}
    
    patient_data = {
        "name": "Api Test Patient",
        "age": 30,
        "gender": "Male",
        "contact_number": "9999999999",
        "address": "Test Avenue",
        "complaint": "Fever",
        "priority": "NORMAL"
    }
    create_res = requests.post(f"{BASE_URL}/patients/", json=patient_data, headers=headers)
    patient = create_res.json()
    patient_id = patient["id"]
    patient_code = patient["patient_code"]
    doctor_id = patient["doctor_id"]
    print(f"✅ Patient Created: {patient_code} assigned to Doctor ID {doctor_id}")
    
    # 2. Doctor logs in and prescribes
    staff_res = requests.get(f"{BASE_URL}/admin/staff", headers=headers)
    staff_list = staff_res.json()
    doctor = next(s for s in staff_list if s["id"] == doctor_id)
    doc_email = doctor["email"]
    
    doc_login = requests.post(f"{BASE_URL}/token", data={"username": doc_email, "password": "doc123"})
    doc_token = doc_login.json()["access_token"]
    doc_headers = {"Authorization": f"Bearer {doc_token}"}
    
    # Consult done
    requests.post(f"{BASE_URL}/patients/{patient_id}/consultation-done", headers=doc_headers)
    # Prescribe Medicine
    requests.post(f"{BASE_URL}/pharmacy/prescribe", json={"patient_id": patient_id, "medicine_name": "Paracetamol", "dosage_instructions": "1-0-1"}, headers=doc_headers)
    # Prescribe Lab Test
    requests.post(f"{BASE_URL}/lab-reports/prescribe", json={"patient_id": patient_id, "test_name": "Blood Test", "notes": "Fever profile"}, headers=doc_headers)
    print("✅ Doctor marked consultation done and prescribed medicine & lab test")
    
    # 3. Nurse assigns lab tech & adds meds (if applicable)
    nurse_res = requests.get(f"{BASE_URL}/staff/my-nurses", headers=doc_headers) # actually doc gets their assigned nurse? Let's just use admin to find a nurse
    nurse = next(s for s in staff_list if s["role"] == "NURSE" and s.get("assigned_doctor_id") == doctor_id)
    nurse_email = nurse["email"]
    
    nurse_login = requests.post(f"{BASE_URL}/token", data={"username": nurse_email, "password": "nurse123"})
    nurse_token = nurse_login.json()["access_token"]
    nurse_headers = {"Authorization": f"Bearer {nurse_token}"}
    
    # Find pending lab test and assign
    pending_tests_res = requests.get(f"{BASE_URL}/lab-reports/?status=PENDING", headers=nurse_headers)
    pending_tests = pending_tests_res.json()
    pt_test = next(t for t in pending_tests if t["patient_id"] == patient_id and t["test_name"] == "Blood Test")
    
    lab_tech = next(s for s in staff_list if s["role"] == "LAB_TECHNICIAN")
    requests.put(f"{BASE_URL}/lab-reports/{pt_test['id']}/assign", json={"technician_id": lab_tech["id"]}, headers=nurse_headers)
    print("✅ Nurse assigned Lab Tech")
    
    # 4. Lab Tech uploads report
    lab_login = requests.post(f"{BASE_URL}/token", data={"username": lab_tech["email"], "password": "lab123"})
    lab_headers = {"Authorization": f"Bearer {lab_login.json()['access_token']}"}
    
    # fake file upload
    with open("dummy.txt", "w") as f: f.write("dummy content")
    with open("dummy.txt", "rb") as f:
        files = {"file": ("dummy.txt", f, "text/plain")}
        requests.post(f"{BASE_URL}/lab-reports/{pt_test['id']}/upload", files=files, headers=lab_headers)
    print("✅ Lab uploaded report")
    
    # 5. Doctor reviews report
    requests.put(f"{BASE_URL}/lab-reports/{pt_test['id']}/review", headers=doc_headers)
    print("✅ Doctor reviewed report")
    
    # 6. Pharmacy dispenses
    pharm_login = requests.post(f"{BASE_URL}/token", data={"username": "pharmacy@medflow.com", "password": "pharmacy123"})
    pharm_headers = {"Authorization": f"Bearer {pharm_login.json()['access_token']}"}
    
    prescriptions_res = requests.get(f"{BASE_URL}/pharmacy/?status=PENDING", headers=pharm_headers)
    presc = next(p for p in prescriptions_res.json() if p["patient_id"] == patient_id)
    requests.put(f"{BASE_URL}/pharmacy/{presc['id']}/dispense", headers=pharm_headers)
    print("✅ Pharmacy dispensed medicine")
    
    # 7. Billing completes payment
    bill_login = requests.post(f"{BASE_URL}/token", data={"username": "billing@medflow.com", "password": "billing123"})
    bill_headers = {"Authorization": f"Bearer {bill_login.json()['access_token']}"}
    
    bills_res = requests.get(f"{BASE_URL}/billing/?status=PENDING", headers=bill_headers)
    bill = next(b for b in bills_res.json() if b["patient_id"] == patient_id)
    requests.put(f"{BASE_URL}/billing/{bill['id']}/pay", headers=bill_headers)
    print("✅ Billing marked payment complete")
    
    # 8. Nurse marks final discharge
    requests.post(f"{BASE_URL}/patients/{patient_id}/discharge", headers=nurse_headers)
    print("✅ Nurse discharged patient")
    
    # Check final patient status
    patient_final = requests.get(f"{BASE_URL}/patients/{patient_id}", headers=admin_headers).json()
    print(f"🏆 TEST COMPLETE. Final Patient Status: {patient_final['status']}")

if __name__ == "__main__":
    try:
        run_test()
    except Exception as e:
        print(f"❌ Error during test: {str(e)}")

File: frontend/test_e2e.js
const fs = require('fs');

const BASE_URL = "http://127.0.0.1:8001";

async function runTest() {
    try {
        console.log("Starting E2E Node API Test against " + BASE_URL);

        // 1. Super Admin creates patient
        const adminLoginRes = await fetch(`${BASE_URL}/token`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: new URLSearchParams({ username: "admin@medflow.com", password: "admin123" })
        });
        if (!adminLoginRes.ok) throw new Error("Admin login failed");
        const adminData = await adminLoginRes.json();
        const adminToken = adminData.access_token;
        const adminHeaders = { 'Authorization': `Bearer ${adminToken}`, 'Content-Type': 'application/json' };

        const patientData = {
            name: "Api Test Patient Node",
            age: 30,
            gender: "Male",
            contact_number: "9999999999",
            address: "Test Avenue",
            complaint: "Fever",
            priority: "NORMAL"
        };
        const createRes = await fetch(`${BASE_URL}/patients/`, {
            method: 'POST',
            body: JSON.stringify(patientData),
            headers: adminHeaders
        });
        if (!createRes.ok) throw new Error(await createRes.text());
        const patient = await createRes.json();
        console.log(`✅ Patient Created: ${patient.patient_code} assigned to Doctor ID ${patient.doctor_id}`);

        // 2. Doctor logs in and prescribes
        const staffRes = await fetch(`${BASE_URL}/admin/staff`, { headers: adminHeaders });
        const staffList = await staffRes.json();
        const doctor = staffList.find(s => s.id === patient.doctor_id);

        const docLogin = await fetch(`${BASE_URL}/token`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: new URLSearchParams({ username: doctor.email, password: "doc123" })
        });
        const docToken = (await docLogin.json()).access_token;
        const docHeaders = { 'Authorization': `Bearer ${docToken}`, 'Content-Type': 'application/json' };

        await fetch(`${BASE_URL}/patients/${patient.id}/consultation-done`, { method: 'POST', headers: docHeaders });
        await fetch(`${BASE_URL}/pharmacy/prescribe`, {
            method: 'POST',
            body: JSON.stringify({ patient_id: patient.id, medicine_name: "Paracetamol", dosage_instructions: "1-0-1" }),
            headers: docHeaders
        });
        await fetch(`${BASE_URL}/lab-reports/prescribe`, {
            method: 'POST',
            body: JSON.stringify({ patient_id: patient.id, test_name: "Blood Test", notes: "Fever" }),
            headers: docHeaders
        });
        console.log("✅ Doctor prescribed medicine & lab test");

        // 3. Nurse
        const nurse = staffList.find(s => s.role === "NURSE"); // just using first nurse
        const nurseLogin = await fetch(`${BASE_URL}/token`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: new URLSearchParams({ username: nurse.email, password: "nurse123" })
        });
        const nurseToken = (await nurseLogin.json()).access_token;
        const nurseHeaders = { 'Authorization': `Bearer ${nurseToken}`, 'Content-Type': 'application/json' };

        const pendingTestsRes = await fetch(`${BASE_URL}/lab-reports/?status=PENDING`, { headers: nurseHeaders });
        const ptTest = (await pendingTestsRes.json()).find(t => t.patient_id === patient.id);

        const labTech = staffList.find(s => s.role === "LAB_TECHNICIAN");
        await fetch(`${BASE_URL}/lab-reports/${ptTest.id}/assign`, {
            method: 'PUT',
            body: JSON.stringify({ technician_id: labTech.id }),
            headers: nurseHeaders
        });
        console.log("✅ Nurse assigned Lab Tech");

        // 4. Lab Tech uploads
        const labLogin = await fetch(`${BASE_URL}/token`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: new URLSearchParams({ username: labTech.email, password: "lab123" })
        });
        const labToken = (await labLogin.json()).access_token;

        // Fake multipart upload
        const boundary = '----WebKitFormBoundary7MA4YWxkTrZu0gW';
        let body = `--${boundary}\r\n`;
        body += `Content-Disposition: form-data; name="file"; filename="dummy.txt"\r\n`;
        body += `Content-Type: text/plain\r\n\r\n`;
        body += `dummy content\r\n`;
        body += `--${boundary}--\r\n`;

        await fetch(`${BASE_URL}/lab-reports/${ptTest.id}/upload`, {
            method: 'POST',
            body: body,
            headers: {
                'Authorization': `Bearer ${labToken}`,
                'Content-Type': `multipart/form-data; boundary=${boundary}`
            }
        });
        console.log("✅ Lab uploaded report");

        // 5. Doctor Reviews
        await fetch(`${BASE_URL}/lab-reports/${ptTest.id}/review`, { method: 'PUT', headers: docHeaders });
        console.log("✅ Doctor reviewed report");

        // 6. Pharmacy
        const pharmLogin = await fetch(`${BASE_URL}/token`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: new URLSearchParams({ username: "pharmacy@medflow.com", password: "pharmacy123" })
        });
        const pharmHeaders = { 'Authorization': `Bearer ${(await pharmLogin.json()).access_token}`, 'Content-Type': 'application/json' };

        const prescRes = await fetch(`${BASE_URL}/pharmacy/?status=PENDING`, { headers: pharmHeaders });
        const presc = (await prescRes.json()).find(p => p.patient_id === patient.id);
        await fetch(`${BASE_URL}/pharmacy/${presc.id}/dispense`, { method: 'PUT', headers: pharmHeaders });
        console.log("✅ Pharmacy dispensed 100%");

        // 7. Billing
        const billLogin = await fetch(`${BASE_URL}/token`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: new URLSearchParams({ username: "billing@medflow.com", password: "billing123" })
        });
        const billHeaders = { 'Authorization': `Bearer ${(await billLogin.json()).access_token}`, 'Content-Type': 'application/json' };

        const billRes = await fetch(`${BASE_URL}/billing/?status=PENDING`, { headers: billHeaders });
        const billList = await billRes.json();
        const bill = billList.find(b => b.patient_id === patient.id);
        if (bill) {
            const payRes = await fetch(`${BASE_URL}/billing/${bill.id}/pay`, { method: 'PUT', headers: billHeaders });
            if (!payRes.ok) console.log("Billing error", await payRes.text());
        }
        console.log("✅ Billing paid");

        // 8. Discharge
        await fetch(`${BASE_URL}/patients/${patient.id}/discharge`, { method: 'POST', headers: nurseHeaders });

        // Fetch Final Patient
        const finalPatient = await fetch(`${BASE_URL}/patients/${patient.id}`, { headers: adminHeaders }).then(r => r.json());
        console.log(`🏆 TEST COMPLETE. Final status: ${finalPatient.status}`);

    } catch (e) {
        console.error("Test failed: ", e.message);
    }
}

runTest();
